src/infer_context.cc

```diff
@@ -46,13 +46,34 @@ InferContext::Init()
   }
 }
 
+void
+InferContext::PrepareRequest()
+{
+  // TPA-560: Explicitly restrict the function to dynamic grpc client backend
+  // only for now. Understand why Triton grpc client doesn't require resetting
+  // the stream across multiple requests.
+  if (backend_kind_ != cb::BackendKind::DYNAMIC_GRPC) {
+    return;
+  }
+
+  // Dynamic grpc client requires to restart the stream before every new request
+  // because the for each request, the stream is half-closed from the client
+  // side due to calling WritesDone.
+  if (streaming_ && inference_started_) {
+    infer_backend_->StopStream();
+    infer_backend_->StartStream(
+        async_callback_func_, (!parser_->IsDecoupled()));
```

Comment:
It looks like PrepareRequest is called for every request sent.

If the `DynamicGrpcClient::StartStream` and `DynamicGrpcClient::StopStream` really need to be called at the start and end of every request, then let's just encapsulate that logic inside of Infer so that we don't need to change more out here.